### PR TITLE
[RELAND2][dynamo][nn-modules] Trace through nn.Module dunder methods for UnspecializedNNModule

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128148
* #128470
* #128463
* #128453
* #128428
* #128315
* #128440
* __->__ #126578

Tracing through `__init__`  is important because it initializes (calls STORE_ATTR) on members. By doing that, we kick in the mutation tracking for these objects. So, things like mutating `_modules` etc is tracked automatically.

Fixes https://github.com/pytorch/pytorch/issues/111837

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/nn_module.py [hunk 1]
```diff
@@ -774,11 +772,16 @@ def unpack_var_sequence(self, tx):
...
         ):
-            assert self.source
-            return [
-                VariableBuilder(tx, source=GetItemSource(self.source, idx))(item)
-                for idx, item in enumerate(self.value)
-            ]
+            # The program can mutate the nn module object but the saved `value`
+            # will not reflect the mutations. So, trace through the `__iter__`
+            # function to reflect any tracked mutations.
+            return tx.inline_user_function_return(
```
Comment (anijain2305) [inline torch/_dynamo/variables/nn_module.py:778]: Trace `__iter__`

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -57,10 +57,7 @@
 
 
 def is_standard_setattr(val):
-    return val in (
-        object.__setattr__,
-        torch.nn.Module.__setattr__,
-    )
+    return val in (object.__setattr__,)
```
Comment (anijain2305) [inline torch/_dynamo/variables/user_defined.py:60]: Trace __setattr__

File: torch/_dynamo/variables/user_defined.py [hunk 3]
```diff
@@ -1001,8 +990,29 @@ def call_hasattr(self, tx, name: str) -> "VariableTracker":
...
+
+        getattr_fn = self._check_for_getattr()
+        if isinstance(getattr_fn, types.FunctionType):
+            # Dynamo is going to trace the __getattr__ function with
+            # args=name. Set the source accordingly.
+            new_source = None
+            if self.source:
+                new_source = AttrSource(self.source, "__getattr__")
+            try:
+                result = variables.UserMethodVariable(
```
Comment (anijain2305) [inline torch/_dynamo/variables/user_defined.py:1004]: Trace `__getattr__`

File: torch/_dynamo/variables/misc.py [hunk 4]
```diff
@@ -175,6 +173,19 @@ def call_method(
...
+        elif inner_fn is object.__delattr__:
+            attr = args[0]
+            try:
+                attr = attr.as_python_constant()
+            except NotImplementedError:
+                unimplemented(f"non-const delattr attr: {attr}")
+            if not tx.output.side_effects.is_attribute_mutation(self.objvar):
+                unimplemented(f"delattr({self.objvar}, {attr}, ...)")
+
+            tx.output.side_effects.store_attr(
```
Comment (anijain2305) [inline torch/_dynamo/variables/misc.py:185]: Trace `__delattr__`

File: torch/_dynamo/variables/misc.py [hunk 5]
```diff
@@ -133,12 +134,9 @@ def call_method(
                 and isinstance(objvar.mutable_local, AttributeMutationNew)
                 and not (args or kwargs)
             ):
-                tx.output.side_effects.store_attr(
-                    objvar,
-                    "__call_nn_module_init",
-                    variables.ConstantVariable.create(True),
-                )
-                return variables.ConstantVariable.create(None)
+                return variables.UserFunctionVariable(
```
Comment (anijain2305) [inline torch/_dynamo/variables/misc.py:137]: Trace `__init__`

File: torch/_dynamo/utils.py [hunk 6]
```diff
@@ -1939,12 +1939,12 @@ def object_has_getattribute(value: Any):
...
 
 
-def get_custom_getattr(value: Any):
+def get_custom_getattr(value: Any, ignore_nn_module_getattr: bool = False):
     try:
         getattr_fn = inspect.getattr_static(type(value), "__getattr__")
     except AttributeError:
         getattr_fn = None
-    if getattr_fn is torch.nn.Module.__getattr__:
+    if ignore_nn_module_getattr and getattr_fn is torch.nn.Module.__getattr__:
```
Comment (anijain2305) [inline torch/_dynamo/utils.py:1947]: For `NNModuleVariable`

**Review by jansel — APPROVED** [review]

**Comment by anijain2305** [comment]
Extra change from reland attempt 1 is [`2d9424a` (#126578)](https://github.com/pytorch/pytorch/pull/126578/commits/2d9424a0c0a1dffcf0ff3f47c2702b2fa91a5871)

**Pull request closed without merging**